File: Examples/python/generate_arbitrary_waveform.py
import redpitaya_scpi as scpi
import sys
import math


rp_s = scpi.scpi(sys.argv[1])

rp_s.tx_txt('ACQ:BUF:SIZE?')
BUFF_SIZE = int(rp_s.rx_txt())

y = ''
x = ''
t = []

for i in range(0, BUFF_SIZE):
	t.append((2 * math.pi) / BUFF_SIZE * i)

for i in range(0, BUFF_SIZE-1):
	if(i != BUFF_SIZE-2): 
		b = math.sin(t[i]) + (1.0/3.0) + math.sin(t[i] * 3)
		if(b <= -1 or b >= 1):
			x += str(-1.0) + ', '
			y += str(-1.0) + ', '
		else: 
			x += str(math.sin(t[i]) + (1.0/3.0) + math.sin(t[i] * 3)) + ', '
			y += str((1.0 / 2.0) * math.sin(t[i]) + (1.0/4.0) * math.sin(t[i] * 4)) + ', '
			
	else:
		c = math.sin(t[i]) + (1.0/3.0) + math.sin(t[i] * 3)
		if(c <= -1 or c >= 1):
			x += str(-1.0)
			y += str(-1.0) 
		else:
			x += str(math.sin(t[i]) + (1.0/3.0) + math.sin(t[i] * 3))
			y += str((1.0 / 2.0) * math.sin(t[i]) + (1.0/4.0) * math.sin(t[i] * 4))		

rp_s.tx_txt('SOUR1:FUNC ARBITRARY')
rp_s.tx_txt('SOUR2:FUNC ARBITRARY')

rp_s.tx_txt('SOUR1:TRAC:DATA:DATA ' + x)
rp_s.tx_txt('SOUR2:TRAC:DATA:DATA ' + y)

rp_s.tx_txt('SOUR1:VOLT 1')
rp_s.tx_txt('SOUR2:VOLT 1')

rp_s.tx_txt('SOUR1:FREQ:FIX 4000')
rp_s.tx_txt('SOUR2:FREQ:FIX 4000')

rp_s.tx_txt('OUTPUT1:STATE ON')
rp_s.tx_txt('OUTPUT2:STATE ON')

#print x

rp_s.tx_txt('SOUR1:TRAC:DATA:DATA?')
buff_string = rp_s.rx_txt()
buff_string = buff_string.strip('{}\n\r').replace("  ", "").split(',')
buff = map(float, buff_string)

import matplotlib.pyplot as plt
plt.plot(buff)
plt.ylabel('Voltage')
plt.show()

File: Examples/Matlab/generate_arbitrary_waveform.m
%% Define Red Pitaya as TCP/IP object
clc
clear all
close all
IP= '192.168.178.102';           % Input IP of your Red Pitaya...
port = 5000;
tcpipObj=tcpip(IP, port);

tcpipObj.InputBufferSize = 16384*64;
tcpipObj.OutputBufferSize = 16384*64;
flushinput(tcpipObj)
flushoutput(tcpipObj)

%% Open connection with your Red Pitaya and close previous
x=instrfind;
fclose(x);
fopen(tcpipObj);
tcpipObj.Terminator = 'CR/LF';

%% Calcualte arbitrary waveform with 16384 samples
% Values of arbitrary waveform must be in range from -1 to 1.
N=16383;
t=0:(2*pi)/N:2*pi;
x=sin(t)+1/3*sin(3*t);
y=1/2*sin(t)+1/4*sin(4*t);
plot(t,x,t,y)
grid on

%% Convert waveforms to string with 5 decimal places accuracy
waveform_ch_1_0 =num2str(x,'%1.5f,');
waveform_ch_2_0 =num2str(y,'%1.5f,');

% latest are empty spaces  “,”.
waveform_ch_1 =waveform_ch_1_0(1,1:length(waveform_ch_1_0)-3);
waveform_ch_2 =waveform_ch_2_0(1,1:length(waveform_ch_2_0)-3);

%%

fprintf(tcpipObj,'GEN:RST')                     % Reset to default settings

fprintf(tcpipObj,'SOUR1:FUNC ARBITRARY');       % Set function of output signal
fprintf(tcpipObj,'SOUR2:FUNC ARBITRARY');       % {sine, square, triangle, sawu, sawd}

fprintf(tcpipObj,['SOUR1:TRAC:DATA:DATA ' waveform_ch_1])  % Send waveforms to Red Pitya
fprintf(tcpipObj,['SOUR2:TRAC:DATA:DATA ' waveform_ch_2])

fprintf(tcpipObj,'SOUR1:VOLT 0.7');             % Set amplitude of output signal
fprintf(tcpipObj,'SOUR2:VOLT 1');

fprintf(tcpipObj,'SOUR1:FREQ:FIX 4000');        % Set frequency of output signal
fprintf(tcpipObj,'SOUR2:FREQ:FIX 4000');


fprintf(tcpipObj,'OUTPUT1:STATE ON');
fprintf(tcpipObj,'OUTPUT2:STATE ON');

fclose(tcpipObj);